pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,309 @@
+	// this contains the matchers which are not related to labelName
+	var unrelatedMatchers []*labels.Matcher
+
+	// this contains the pos of labelValues that are matched by the matchers
+	var matchedValuesPos = make([]int, len(labelValues))
+	for pos := range labelValues {
+		matchedValuesPos[pos] = pos
+	}
+
+	for _, m := range matchers {
+		if m.Name != labelName {
+			// check if has the retained label name
+			if m.Name == retainExistingPrefix+labelName {
+				unrelatedMatchers = append(unrelatedMatchers, &labels.Matcher{
+					Name:  labelName,
+					Type:  m.Type,
+					Value: m.Value,
+				})
+			} else {
+				unrelatedMatchers = append(unrelatedMatchers, m)
+			}
+			continue
+		}
+
+		var matchedValuesPosThisMatcher []int
+		for _, v := range matchedValuesPos {
+			if m.Matches(labelValues[v]) {
+				matchedValuesPosThisMatcher = append(matchedValuesPosThisMatcher, v)
+			}
+		}
+		matchedValuesPos = matchedValuesPosThisMatcher
+	}
+	return matchedValuesPos, unrelatedMatchers
+}
+
+type addLabelsSeriesSet struct {
+	upstream storage.SeriesSet
+	labels   labels.Labels
+}
+
+func (m *addLabelsSeriesSet) Next() bool {
+	return m.upstream.Next()
+}
+
+// At returns full series. Returned series should be iteratable even after Next is called.
+func (m *addLabelsSeriesSet) At() storage.Series {
+	return &addLabelsSeries{
+		upstream: m.upstream.At(),
+		labels:   m.labels,
+	}
+}
+
+// The error that iteration as failed with.
+// When an error occurs, set cannot continue to iterate.
+func (m *addLabelsSeriesSet) Err() error {
+	return m.upstream.Err()
+}
+
+// A collection of warnings for the whole set.
+// Warnings could be return even iteration has not failed with error.
+func (m *addLabelsSeriesSet) Warnings() storage.Warnings {
+	return m.upstream.Warnings()
+}
+
+type addLabelsSeries struct {
+	upstream storage.Series
+	labels   labels.Labels
+}
+
+// Labels returns the complete set of labels. For series it means all labels identifying the series.
+func (a *addLabelsSeries) Labels() labels.Labels {
+	return setLabelsRetainExisting(a.upstream.Labels(), a.labels...)
+}
+
+// Iterator returns a new, independent iterator of the data of the series.
+func (a *addLabelsSeries) Iterator() chunkenc.Iterator {
+	return a.upstream.Iterator()
+}
+
+// this sets a label and preserves an existing value a new label prefixed with
+// original_. It doesn't do this recursively.
+func setLabelRetainExisting(ls labels.Labels, lb *labels.Builder, additional labels.Label) {
+	if oldValue := ls.Get(additional.Name); oldValue != "" {
+		lb.Set(
+			retainExistingPrefix+additional.Name,
+			oldValue,
+		)
+	}
+	lb.Set(additional.Name, additional.Value)
+}
+
+func setLabelsRetainExisting(src labels.Labels, additionalLabels ...labels.Label) labels.Labels {
+	lb := labels.NewBuilder(src)
+
+	for _, additionalL := range additionalLabels {
+		setLabelRetainExisting(src, lb, additionalL)
```

Comment:
nit: I would suggest to inline this call. `setLabelRetainExisting` is not called anywhere else, and this method is somewhat empty.

Reply:
Inlined this per 6dde8c29a8827e0d0deb3bb32e809a846d37d932